pkg/controller/endpointslice/endpointslice_controller.go

```diff
@@ -313,6 +313,15 @@ func (c *Controller) syncService(key string) error {
 		return err
 	}
 
+	if service.DeletionTimestamp != nil {
```

Comment:
I'm trying to think through what this means if the service has a custom finalizer and deletionTimestamp, and the custom finalizer takes noticeable time to remove. Returning here means any existing endpointslices will go stale, could contain addresses that don't exist any more, won't pick up new addresses.

What does kube-proxy do with services with a deletionTimestamp?

Reply:
That's a great question. As far as I can tell, kube-proxy does not have any special treatment for Services with a deletionTimestamp specified. I think once a Service has begun the deletion process the primary goal is to close out existing connections, something that this would still enable. I can't think of a great way to not potentially conflict with garbage collection while also preventing stale EndpointSlices during the deletion process. 

@freehan do you have any ideas about what should be done during Service deletion? Trying to understand what we should do for EndpointSlices when Services with finalizers are being deleted. Is it acceptable to not change EndpointSlices at all once the deletion process begins?

Reply:
Might be worth changing kube-proxy to ignore service with DeletionTimestamp. 

I think it is totally okay for endpointSlice Controller to stop managing EndpointSlices of Service with DeletionTimestamp. And leave it to K8s GarbageCollector to remove EndpointSlice objects once done.

Reply:
> I can't think of a great way to not potentially conflict with garbage collection while also preventing stale EndpointSlices during the deletion process.

An alternative is to modify `reconciler#finalize` to skip creations if the service's deletionTimestamp is set, but to permit updates/deletes. That ensures existing endpointslices stay updated and that we prune stale endpoints, but that we don't create new EndpointSlice objects for a service being deleted.

Reply:
Good idea - I've switched to that approach.

Reply:
@freehan I am a bit torn on the best approach here. In some sense, updating kube-proxy to consider Services with a deletionTimestamp deleted/gone would result in the most consistent user experience. On the other hand, that does feel relatively harsh, especially if there are finalizers that rely on connectivity continuing for some period of time. Although this will result in a potentially inconsistent experience, it seems like it could be better and easier to understand than dropping everything immediately. Admittedly I'm not sure how much people care about connections on Services that are being deleted.